contracts/marketpools.js

```diff
@@ -0,0 +1,348 @@
+      quoteQuantity: 0,
+      quoteVolume: 0,
+      quotePrice: 0, // Quote per Base (reverse price)
+      precision: Math.min(baseToken.precision, quoteToken.precision),
+      active: true,
+      creator: api.sender,
+    };
+    await api.db.insert('pools', newPool);
+
+    // burn the token creation fees
+    if (api.sender !== api.owner && api.BigNumber(poolCreationFee).gt(0)) {
+      await api.executeSmartContract('tokens', 'transfer', {
+        // eslint-disable-next-line no-template-curly-in-string
+        to: 'null', symbol: "'${CONSTANTS.UTILITY_TOKEN_SYMBOL}$'", quantity: poolCreationFee, isSignedWithActiveKey,
+      });
+    }
+    api.emit('createPool', { tokenPair });
+  }
+};
+
+actions.addLiquidity = async (payload) => {
+  const {
+    tokenPair,
+    baseQuantity,
+    quoteQuantity,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (!api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')
+    || !api.assert(baseQuantity && api.BigNumber(baseQuantity).gt(0), 'invalid baseQuantity')
+    || !api.assert(quoteQuantity && api.BigNumber(quoteQuantity).gt(0), 'invalid quoteQuantity')
+    || !await validateTokenPair(tokenPair)) {
+    return;
+  }
+
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  const pool = await api.db.findOne('pools', { tokenPair });
+  if (api.assert(pool, 'no existing pool for tokenPair')) {
+    if (api.BigNumber(pool.baseQuantity).eq(0) && api.BigNumber(pool.quoteQuantity).eq(0)
+      && await validateOracle(pool, api.BigNumber(quoteQuantity).dividedBy(baseQuantity)) === false) return;
+    if (api.BigNumber(pool.baseQuantity).gt(0) && api.BigNumber(pool.quoteQuantity).gt(0)
+      && !validateLiquiditySwap(pool, baseQuantity, quoteQuantity)) return;
+
+    // update liquidity position
+    const lp = await api.db.findOne('liquidityPosition', { account: api.sender, tokenPair });
+    if (lp) {
+      lp.baseQuantity = api.BigNumber(lp.baseQuantity).plus(baseQuantity);
+      lp.quoteQuantity = api.BigNumber(lp.quoteQuantity).plus(quoteQuantity);
+      await api.db.update('liquidityPosition', lp);
+    } else {
+      const newlp = {
+        account: api.sender,
+        tokenPair,
+        baseQuantity,
+        quoteQuantity,
+      };
+      await api.db.insert('liquidityPosition', newlp);
+    }
+
+    // deposit requested tokens to contract
+    const baseRes = await api.executeSmartContract('tokens', 'transferToContract', { symbol: baseSymbol, quantity: baseQuantity, to: 'marketpools' });
+    const quoteRes = await api.executeSmartContract('tokens', 'transferToContract', { symbol: quoteSymbol, quantity: quoteQuantity, to: 'marketpools' });
+    if (!api.assert(baseRes.errors === undefined && quoteRes.errors === undefined, 'deposit transfer errors')) return;
+    updatePoolStats(pool, baseQuantity, quoteQuantity, false);
+    api.emit('addLiquidity', { memo: `Add ${baseSymbol} and ${quoteSymbol}` });
```

Comment:
Perhaps the emission should also contain the quantities added?

Reply:
Should be fine as-is since quantities will be recorded in the transferToContract events.